numpy/core/src/umath/ufunc_type_resolution.c

```diff
@@ -158,12 +160,28 @@ raise_casting_error(
         return -1;
     }
 
+    int j, nin = ufunc->nin, nop = nin + ufunc->nout;
+    PyObject *froms = PyTuple_New(nop);
+    PyObject *tos = PyTuple_New(nop);
+
+    if (input) {
+        for(j=0;j<nop;j++){
+            PyTuple_SetItem(froms, j, (PyObject *)PyArray_DESCR(operands[j]));
+            PyTuple_SetItem(tos, j, (PyObject *)dtypes[j]);
```

Comment:
Missing `Py_INCREF(value)` here - `PyTuple_SetItem` steals references.

Also, you might as well use `PyTuple_SETITEM`, it's faster and no more dangerous

Reply:
Thanks for all your helpful comments so far!
I included the `Py_INCREF` on the `dtypes[j]` object and it removed a segfault in an edge case I found. I don't think there needs to be one with regards to the `PyArray_DESCR(operands[j])` object because `PyArray_DESRC` returns a borrowed reference, correct me if i misunderstand something there.

Reply:
Both are borrowed references, and this is exactly why you do need to incref them an extra time - when you call a function like `SETITEM` that steals a reference, you must make sure the reference is yours to give (ie not borrowed).